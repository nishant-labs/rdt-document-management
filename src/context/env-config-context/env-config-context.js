import React, { useEffect, useState } from 'react';
import { fetchEnvConfig } from '../../utils/http';

export const EnvConfigContext = React.createContext();

export const EnvConfigContextProvider = ({ children }) => {
  const [envConfigData, setEnvConfigData] = useState();
  const [error, setError] = useState();

  useEffect(() => {
    fetchEnvConfig().then(({ error, payload }) => {
      setEnvConfigData(payload);
      setError(error);
    });
  }, []);

  return (
    <EnvConfigContext.Provider value={{ envConfigData, error }}>
      {envConfigData ? children : <div>Loading app...</div>}
      {error && <div>Something gone wrong while fetching config</div>}
    </EnvConfigContext.Provider>
  );
};
