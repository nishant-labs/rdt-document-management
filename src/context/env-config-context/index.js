export * from './env-config-context';
