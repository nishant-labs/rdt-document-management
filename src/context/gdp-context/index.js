export * from './gdp-context';
export * from './selectors';
