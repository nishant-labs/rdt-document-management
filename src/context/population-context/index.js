export * from './population-context';
export * from './selectors';
