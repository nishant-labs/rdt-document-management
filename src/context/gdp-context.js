import React, { useCallback, useContext, useState } from 'react';
import { fetchGDPFromWorldBank } from '../utils/http';
import { EnvConfigContext } from './env-config-context';

export const GdpContext = React.createContext();

export const GdpContextProvider = ({ children }) => {
	const { envConfigData } = useContext(EnvConfigContext);
	const [gdpData, setGdpData] = useState();
	const [gdpApiError, setError] = useState();

	const fetchGdp = useCallback(
		async (fullYear) => {
			const { error, payload } = await fetchGDPFromWorldBank(
				envConfigData.baseUrl,
				fullYear
			);
			setGdpData(payload);
			setError(error);
		},
		[envConfigData]
	);

	return (
		<GdpContext.Provider value={{ gdpData, gdpApiError, fetchGdp }}>
			{children}
		</GdpContext.Provider>
	);
};
