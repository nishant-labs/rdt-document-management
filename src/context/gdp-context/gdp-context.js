import React, { useCallback, useContext, useState } from 'react';
import { fetchGDPFromWorldBank } from '../../utils/http';
import { EnvConfigContext } from '../env-config-context';

export const GdpContext = React.createContext();
export const GdpContextConsumer = GdpContext.Consumer;

export const GdpContextProvider = ({ children }) => {
	const { envConfigData } = useContext(EnvConfigContext);
	const [gdpData, setGdpData] = useState();
	const [gdpApiError, setGdpApiError] = useState();

	const fetchGdp = useCallback(
		async (fullYear) => {
			const { error, payload } = await fetchGDPFromWorldBank(
				envConfigData.baseUrl,
				fullYear
			);
			setGdpData(payload);
			setGdpApiError(error);
		},
		[envConfigData]
	);

	const resetGdpData = useCallback(() => {
		setGdpData(undefined);
		setGdpApiError(undefined);
	}, []);

	return (
		<GdpContext.Provider
			value={{ gdpData, gdpApiError, fetchGdp, resetGdpData }}
		>
			{children}
		</GdpContext.Provider>
	);
};
