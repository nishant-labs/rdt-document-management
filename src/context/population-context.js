import React, { useCallback, useContext, useState } from 'react';
import { fetchPopulationFromWorldBank } from '../utils/http';
import { EnvConfigContext } from './env-config-context';

export const PopulationContext = React.createContext();

export const PopulationContextProvider = ({ children }) => {
	const { envConfigData } = useContext(EnvConfigContext);
	const [populationData, setPopulationData] = useState();
	const [error, setError] = useState();

	const fetchPopulation = useCallback(
		async (fullYear) => {
			const { error, payload } = await fetchPopulationFromWorldBank(
				envConfigData.baseUrl,
				fullYear
			);
			setPopulationData(payload);
			setError(error);
		},
		[envConfigData]
	);

	return (
		<PopulationContext.Provider
			value={{ populationData, error, fetchPopulation }}
		>
			{children}
		</PopulationContext.Provider>
	);
};
