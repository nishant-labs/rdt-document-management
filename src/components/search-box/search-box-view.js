import { useCallback, useContext, useState } from 'react';
import Button from '@material-ui/core/Button';
import TextField from '@material-ui/core/TextField';
import Box from '@material-ui/core/Box';
import { GdpContext } from '../../context/gdp-context';
import { PopulationContext } from '../../context/population-context';

export const SearchBoxView = () => {
	const { fetchGdp } = useContext(GdpContext);
	const { fetchPopulation } = useContext(PopulationContext);
	const [hasError, setHasError] = useState(false);
	const [searchValue, setSearchValue] = useState('');
	const [isTouched, setIsTouched] = useState(false);

	const handleBlur = useCallback(() => {
		setIsTouched(true);
	}, []);

	const handleChange = useCallback(({ target: { value } }) => {
		const isInValid = value.length < 4 || isNaN(value, 10);
		setHasError(isInValid);
		setSearchValue(value);
	}, []);

	const handleSearchClick = useCallback(() => {
		fetchGdp(searchValue);
		fetchPopulation(searchValue);
	}, [fetchGdp, fetchPopulation, searchValue]);

	const showError = hasError && isTouched;
	const isButtonDisabled = hasError || !(isTouched && !!searchValue);
	return (
		<Box display="flex" flexDirection="row" justifyContent="center" p={1} m={1}>
			<Box p={1}>
				<TextField
					error={showError}
					value={searchValue}
					onBlur={handleBlur}
					onChange={handleChange}
					variant="outlined"
					label="Enter Year"
					helperText={showError && 'Enter 4 digit full year, e.g. 2019, 1997 etc...'}
					fullWidth
					required
				/>
			</Box>
			<Box p={2}>
				<Button variant="contained" color="primary" disabled={isButtonDisabled} onClick={handleSearchClick}>
					Get statistics
				</Button>
			</Box>
		</Box>
	);
};
