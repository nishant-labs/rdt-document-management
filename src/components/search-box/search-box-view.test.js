import React from 'react';
import { shallow } from 'enzyme';
import { SearchBoxView } from './search-box-view';
import { PopulationContext } from '../../context/population-context';
import { GdpContext } from '../../context/gdp-context';
import { EnvConfigContext } from '../../context/env-config-context';

const mockGdpContext = {
  fetchGdp: jest.fn(),
  resetGdpData: jest.fn(),
};

const mockPopulationContext = {
  fetchGdp: jest.fn(),
  resetGdpData: jest.fn(),
};

const mockEnvContext = {
  envConfigData: { baseUrl: 'http://example.com' },
  error: null,
};

const setupTest = (setupProps = {}) => {
  const wrapper = shallow(
    <EnvConfigContext.Provider value={mockEnvContext}>
      <PopulationContext.Provider value={mockGdpContext}>
        <GdpContext.Provider value={mockPopulationContext}>
          <SearchBoxView />
        </GdpContext.Provider>
      </PopulationContext.Provider>
    </EnvConfigContext.Provider>
  );
  return {
    wrapper,
    wrapperComponent: wrapper.find(SearchBoxView),
  };
};

describe('components/search-box/search-box-view', () => {
  test('should render', () => {
    const { wrapper } = setupTest();
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
