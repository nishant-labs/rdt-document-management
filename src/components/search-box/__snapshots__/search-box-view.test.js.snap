// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/search-box/search-box-view should render 1`] = `
"<ContextProvider value={{...}}>
  <ContextProvider value={{...}}>
    <SearchBoxView />
  </ContextProvider>
</ContextProvider>"
`;
